sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -783,6 +783,16 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ds.filter(_.b > 1).collect().toSeq
     }
   }
+
+  test("transformed dataset correctly solve the attributes") {
```

Comment:
Thanks a lot @cloud-fan for your review. I think the issue still exists, problem is that unit test actually doesn't reflect this issue, let me change the unit test.

You could try the below to see the issue.

```
val dataset = Seq(1, 2, 3).toDS
val mappedDataset = dataset.map(_ + 1)
mappedDataset.as("t1").joinWith(mappedDataset.as("t2"), $"t1.value" === $"t2.value").show()
```